File: tutorials/04-messaging/python/messaging_demo.py
"""
Build the whole order pipeline with boto3: a topic, two queues, a filter on one
of them, and a dead letter queue.

    pip install -r requirements.txt
    python messaging_demo.py
"""

import json
import os
import time

import boto3

# ---------------------------------------------------------------------------
# The ONLY Floci-specific lines. Delete endpoint_url and this runs on real AWS.
# ---------------------------------------------------------------------------
ENDPOINT = os.environ.get("AWS_ENDPOINT_URL", "http://localhost:4566")
COMMON = dict(
    endpoint_url=ENDPOINT,
    region_name=os.environ.get("AWS_DEFAULT_REGION", "us-east-1"),
    aws_access_key_id=os.environ.get("AWS_ACCESS_KEY_ID", "test"),
    aws_secret_access_key=os.environ.get("AWS_SECRET_ACCESS_KEY", "test"),
)

sqs = boto3.client("sqs", **COMMON)
sns = boto3.client("sns", **COMMON)

SUFFIX = "demo"
EMAIL_Q = f"email-queue-{SUFFIX}"
FRAUD_Q = f"fraud-queue-{SUFFIX}"
DLQ = f"orders-dlq-{SUFFIX}"
MAIN_Q = f"orders-main-{SUFFIX}"
TOPIC = f"orders-{SUFFIX}"


def step(msg):
    print(f"\n==> {msg}")


def make_queue(name, attributes=None):
    url = sqs.create_queue(QueueName=name, Attributes=attributes or {})["QueueUrl"]
    arn = sqs.get_queue_attributes(QueueUrl=url, AttributeNames=["QueueArn"])[
        "Attributes"
    ]["QueueArn"]
    return url, arn


def drain(url, label, wait=5):
    """Read everything currently visible and delete it properly."""
    msgs = sqs.receive_message(
        QueueUrl=url, MaxNumberOfMessages=10, WaitTimeSeconds=wait
    ).get("Messages", [])
    for m in msgs:
        print(f"    {label}: {m['Body'][:80]}")
        # Deleting is a separate call. Receiving only hid the message.
        sqs.delete_message(QueueUrl=url, ReceiptHandle=m["ReceiptHandle"])
    if not msgs:
        print(f"    {label}: nothing")
    return msgs


def main():
    step("Creating queues and topic")
    email_url, email_arn = make_queue(EMAIL_Q)
    fraud_url, fraud_arn = make_queue(FRAUD_Q)
    topic_arn = sns.create_topic(Name=TOPIC)["TopicArn"]
    print(f"    topic: {topic_arn}")

    step("Subscribing both queues to the topic")
    email_sub = sns.subscribe(
        TopicArn=topic_arn, Protocol="sqs", Endpoint=email_arn
    )["SubscriptionArn"]
    fraud_sub = sns.subscribe(
        TopicArn=topic_arn, Protocol="sqs", Endpoint=fraud_arn
    )["SubscriptionArn"]

    # Raw delivery removes the SNS envelope. This is a property of the
    # subscription, so each subscriber chooses independently.
    for sub in (email_sub, fraud_sub):
        sns.set_subscription_attributes(
            SubscriptionArn=sub, AttributeName="RawMessageDelivery", AttributeValue="true"
        )

    # The fraud queue only wants high value orders. SNS applies this before
    # delivery, so the fraud worker never even sees the others.
    sns.set_subscription_attributes(
        SubscriptionArn=fraud_sub,
        AttributeName="FilterPolicy",
        AttributeValue=json.dumps({"value": ["high"]}),
    )
    print("    fraud queue filters on value=high")
    time.sleep(1)

    step("Publishing two orders, one high value and one low")
    for body, value in (("big order", "high"), ("small order", "low")):
        sns.publish(
            TopicArn=topic_arn,
            Message=body,
            MessageAttributes={"value": {"DataType": "String", "StringValue": value}},
        )
    time.sleep(4)

    step("What each queue actually received")
    drain(email_url, "email")
    drain(fraud_url, "fraud")

    step("Dead letter queue: parking a message that never succeeds")
    dlq_url, dlq_arn = make_queue(DLQ)
    main_url, _ = make_queue(
        MAIN_Q,
        {
            "RedrivePolicy": json.dumps(
                {"deadLetterTargetArn": dlq_arn, "maxReceiveCount": "2"}
            ),
            "VisibilityTimeout": "1",
        },
    )
    sqs.send_message(QueueUrl=main_url, MessageBody="poison")

    # Receive it repeatedly without ever deleting it, exactly as a failing
    # worker would. After maxReceiveCount attempts SQS moves it aside.
    for attempt in range(1, 4):
        got = sqs.receive_message(QueueUrl=main_url).get("Messages", [])
        print(f"    attempt {attempt}: {'received' if got else 'nothing visible'}")
        time.sleep(2)

    parked = sqs.get_queue_attributes(
        QueueUrl=dlq_url, AttributeNames=["ApproximateNumberOfMessages"]
    )["Attributes"]["ApproximateNumberOfMessages"]
    print(f"    messages now in the dead letter queue: {parked}")

    step("Cleaning up")
    for url in (email_url, fraud_url, dlq_url, main_url):
        sqs.delete_queue(QueueUrl=url)
    sns.delete_topic(TopicArn=topic_arn)
    print("    done")


if __name__ == "__main__":
    main()

File: tutorials/04-messaging/node/messaging-demo.mjs
/**
 * The same order pipeline as ../python/messaging_demo.py, using AWS SDK v3.
 *
 *   npm install
 *   node messaging-demo.mjs
 */

import {
  SQSClient,
  CreateQueueCommand,
  GetQueueAttributesCommand,
  SendMessageCommand,
  ReceiveMessageCommand,
  DeleteMessageCommand,
  DeleteQueueCommand,
} from "@aws-sdk/client-sqs";
import {
  SNSClient,
  CreateTopicCommand,
  SubscribeCommand,
  SetSubscriptionAttributesCommand,
  PublishCommand,
  DeleteTopicCommand,
} from "@aws-sdk/client-sns";

// ---------------------------------------------------------------------------
// The ONLY Floci-specific lines. Drop `endpoint` and this runs on real AWS.
// ---------------------------------------------------------------------------
const config = {
  endpoint: process.env.AWS_ENDPOINT_URL ?? "http://localhost:4566",
  region: process.env.AWS_DEFAULT_REGION ?? "us-east-1",
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID ?? "test",
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY ?? "test",
  },
};

const sqs = new SQSClient(config);
const sns = new SNSClient(config);

const SUFFIX = "demo-node";
const step = (msg) => console.log(`\n==> ${msg}`);
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function makeQueue(QueueName, Attributes = {}) {
  const { QueueUrl } = await sqs.send(new CreateQueueCommand({ QueueName, Attributes }));
  const { Attributes: a } = await sqs.send(
    new GetQueueAttributesCommand({ QueueUrl, AttributeNames: ["QueueArn"] })
  );
  return { url: QueueUrl, arn: a.QueueArn };
}

async function drain(QueueUrl, label, WaitTimeSeconds = 5) {
  const { Messages = [] } = await sqs.send(
    new ReceiveMessageCommand({ QueueUrl, MaxNumberOfMessages: 10, WaitTimeSeconds })
  );
  for (const m of Messages) {
    console.log(`    ${label}: ${m.Body.slice(0, 80)}`);
    // Deleting is a separate call. Receiving only hid the message.
    await sqs.send(new DeleteMessageCommand({ QueueUrl, ReceiptHandle: m.ReceiptHandle }));
  }
  if (Messages.length === 0) console.log(`    ${label}: nothing`);
  return Messages;
}

async function main() {
  step("Creating queues and topic");
  const email = await makeQueue(`email-queue-${SUFFIX}`);
  const fraud = await makeQueue(`fraud-queue-${SUFFIX}`);
  const { TopicArn } = await sns.send(new CreateTopicCommand({ Name: `orders-${SUFFIX}` }));
  console.log(`    topic: ${TopicArn}`);

  step("Subscribing both queues to the topic");
  const emailSub = (
    await sns.send(new SubscribeCommand({ TopicArn, Protocol: "sqs", Endpoint: email.arn }))
  ).SubscriptionArn;
  const fraudSub = (
    await sns.send(new SubscribeCommand({ TopicArn, Protocol: "sqs", Endpoint: fraud.arn }))
  ).SubscriptionArn;

  // Raw delivery removes the SNS envelope. It is a property of the
  // subscription, so each subscriber chooses independently.
  for (const SubscriptionArn of [emailSub, fraudSub]) {
    await sns.send(
      new SetSubscriptionAttributesCommand({
        SubscriptionArn,
        AttributeName: "RawMessageDelivery",
        AttributeValue: "true",
      })
    );
  }

  // The fraud queue only wants high value orders. SNS applies this before
  // delivery, so the fraud worker never even sees the others.
  await sns.send(
    new SetSubscriptionAttributesCommand({
      SubscriptionArn: fraudSub,
      AttributeName: "FilterPolicy",
      AttributeValue: JSON.stringify({ value: ["high"] }),
    })
  );
  console.log("    fraud queue filters on value=high");
  await sleep(1000);

  step("Publishing two orders, one high value and one low");
  for (const [Message, value] of [
    ["big order", "high"],
    ["small order", "low"],
  ]) {
    await sns.send(
      new PublishCommand({
        TopicArn,
        Message,
        MessageAttributes: { value: { DataType: "String", StringValue: value } },
      })
    );
  }
  await sleep(4000);

  step("What each queue actually received");
  await drain(email.url, "email");
  await drain(fraud.url, "fraud");

  step("Dead letter queue: parking a message that never succeeds");
  const dlq = await makeQueue(`orders-dlq-${SUFFIX}`);
  const main_ = await makeQueue(`orders-main-${SUFFIX}`, {
    RedrivePolicy: JSON.stringify({
      deadLetterTargetArn: dlq.arn,
      maxReceiveCount: "2",
    }),
    VisibilityTimeout: "1",
  });
  await sqs.send(new SendMessageCommand({ QueueUrl: main_.url, MessageBody: "poison" }));

  // Receive it repeatedly without ever deleting it, exactly as a failing
  // worker would. After maxReceiveCount attempts SQS moves it aside.
  for (let attempt = 1; attempt <= 3; attempt++) {
    const { Messages = [] } = await sqs.send(
      new ReceiveMessageCommand({ QueueUrl: main_.url })
    );
    console.log(`    attempt ${attempt}: ${Messages.length ? "received" : "nothing visible"}`);
    await sleep(2000);
  }

  const { Attributes } = await sqs.send(
    new GetQueueAttributesCommand({
      QueueUrl: dlq.url,
      AttributeNames: ["ApproximateNumberOfMessages"],
    })
  );
  console.log(`    messages now in the dead letter queue: ${Attributes.ApproximateNumberOfMessages}`);

  step("Cleaning up");
  for (const QueueUrl of [email.url, fraud.url, dlq.url, main_.url]) {
    await sqs.send(new DeleteQueueCommand({ QueueUrl }));
  }
  await sns.send(new DeleteTopicCommand({ TopicArn }));
  console.log("    done");
}

main().catch((err) => {
  console.error("\nFailed:", err.message);
  process.exit(1);
});
